### PR TITLE
Fix #4882: Make UUID.compareTo() consistent with the JVM.

### LINKED ISSUE #4882: UUID comparison doesn't match JVM
The compare implementation on UUID produces incoherent results.

The following test will pass on the JVM but fail on JS

```scala
import munit.FunSuite

import java.util.UUID

class UUIDSuite extends FunSuite {

  test("v0 UUID should be less than v1 UUID") {
    val v0 = UUID.fromString("1ee23e57-7fff-6490-8018-c46dfbd6cdf7")
    val v1 = UUID.fromString("1ee23e57-8001-6ba0-8000-824f2f22b333")

    assertEquals(v0.compareTo(v1), -1)
  }

}

```

Seems that the culprit is the usage of signed integer arithmetic in :  

https://github.com/scala-js/scala-js/blob/0aabf899e71dc4d31b1a7ca7e205ad7f06c5b41b/javalib/src/main/scala/java/util/UUID.scala#L118-L132  

For the example above, i2 is negative for v1. The condition at L122 returns 1. An unsigned comparison should be used instead.

### PR DESCRIPTION
**Pull request by sjrd**
Because the actual algorithm is unknown and the JavaDoc is not very clear, we have inferred an implementation from trial and error.

### CODE REVIEW

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala [hunk 1]
```diff
@@ -95,21 +95,86 @@ class UUIDTest {
...
+    val sortedUUIDs = uuids.sortWith(referenceLessThan(_, _)).toArray
+
+    /* For reference: full loop to run on the JVM to test all 1M pairs
+     * for (u1 <- sortedUUIDs; u2 <- sortedUUIDs) {
+     *   if (referenceLessThan(u1, u2) != (u1.compareTo(u2) < 0))
+     *     println(s"$u1 $u2")
+     * }
+     */
+
+    // For our unit tests, only test consecutive UUIDs, and assume that transitivy holds
```
Comment (ekrich) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala:170]:
```suggestion
    // For our unit tests, only test consecutive UUIDs, and assume that transitivity holds
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala [hunk 2]
```diff
@@ -95,21 +95,86 @@ class UUIDTest {
...
+    /* #4882 `UUID.compareTo()` is known not to match the specification in RFC
+     * 4122. However, the exact algorithm used by the JVM is not publicly
+     * available with a license that we can use. The best we have is the
+     * JavaDoc that says
+     *
+     * > The first of two UUIDs is greater than the second if the most
+     * > significant field in which the UUIDs differ is greater for the first
+     * > UUID.
+     *
+     * We do not know what a "field" is, but we can guess that it is either
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala:107]: This is a bit misleading IMO. The JavaDoc clearly explains what fields are supposed to be, at least for variant 2:

> The layout of a variant 2 (Leach-Salz) UUID is as follows: The most significant long consists of the following unsigned fields:
> 
>  0xFFFFFFFF00000000 time_low
>  0x00000000FFFF0000 time_mid
>  0x000000000000F000 version
>  0x0000000000000FFF time_hi
>  
> The least significant long consists of the following unsigned fields:
>  0xC000000000000000 variant
>  0x3FFF000000000000 clock_seq
>  0x0000FFFFFFFFFFFF node

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala [hunk 3]
```diff
@@ -95,21 +95,86 @@ class UUIDTest {
...
+     * the pair `get{Least,Most}SignificantBits()`, or the dash-separated
+     * segments of the string representation of a UUID. The latter is of the
+     * form
+     *
+     *   xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+     *      8       4    4    4       12
+     *
+     * Note that the decomposition into `get{Least,Most}SignificantBits()` is
+     * superseded by the decomposition into those segments.
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala:116]: I do not understand what you mean by "superseded" here :-/

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala [hunk 4]
```diff
@@ -95,21 +95,86 @@ class UUIDTest {
...
+
+    /* For reference: full loop to run on the JVM to test all 1M pairs
+     * for (u1 <- sortedUUIDs; u2 <- sortedUUIDs) {
+     *   if (referenceLessThan(u1, u2) != (u1.compareTo(u2) < 0))
+     *     println(s"$u1 $u2")
+     * }
+     */
+
+    // For our unit tests, only test consecutive UUIDs, and assume that transitivity holds
+    for (i <- 0 until (sortedUUIDs.length - 1)) {
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala:171]: Use a zip?

```scala
for {
  (smaller, larger) <- sortedUUIDs.zip(sortedUUIDs.tail)
} {

}
```

**Review by gzm0 — CHANGES_REQUESTED** [review]
Only minor things. I think overall we can & should go with this.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala (hunk 2 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala:107]: If those were indeed the fields used in the comparison, the reference comparison would not have worked for a number of the generated pairs. So they must be talking about different fields.

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/UUIDTest.scala]: I hopefully made it clearer.

**Review by gzm0 — APPROVED** [review]
Much clearer! TY.

**Pull request merged**